instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/ui/routeline/AlternativeRouteSelectionTest.kt

```diff
@@ -0,0 +1,206 @@
+import com.mapbox.navigation.ui.maps.route.line.api.MapboxRouteLineApi
+import com.mapbox.navigation.ui.maps.route.line.api.MapboxRouteLineView
+import com.mapbox.navigation.ui.maps.route.line.model.MapboxRouteLineOptions
+import com.mapbox.navigation.ui.maps.route.line.model.RouteLine
+import kotlinx.android.synthetic.main.activity_basic_navigation_view.*
+import org.junit.Assert.assertEquals
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+
+class AlternativeRouteSelectionTest : BaseTest<BasicNavigationViewActivity>(
+    BasicNavigationViewActivity::class.java
+) {
+
+    @get:Rule
+    val mapboxNavigationRule = MapboxNavigationRule()
+
+    @get:Rule
+    val mockLocationReplayerRule = MockLocationReplayerRule(mockLocationUpdatesRule)
+
+    private lateinit var initIdlingResource: MapStyleInitIdlingResource
+
+    protected lateinit var directionsResponse: DirectionsResponse
+
+    protected lateinit var mapboxNavigation: MapboxNavigation
+
+    protected lateinit var routeLineApi: MapboxRouteLineApi
+    protected lateinit var routeLineView: MapboxRouteLineView
+
+    protected lateinit var mapboxNavigationViewportDataSource: MapboxNavigationViewportDataSource
+    protected lateinit var navigationCamera: NavigationCamera
+
+    protected lateinit var navigationLocationProvider: NavigationLocationProvider
+    protected lateinit var locationPlugin: LocationComponentPlugin
+
+    @Before
+    fun setUp() {
+        initIdlingResource = MapStyleInitIdlingResource(activity.mapView)
+        initIdlingResource.register()
+        Espresso.onIdle()
+
+        directionsResponse = getRoute(activity)
+        val coordinates = directionsResponse.waypoints()?.mapNotNull { it.location() } ?: listOf()
+        val route = directionsResponse.routes()[0]
+
+        runOnMainSync {
+            mockLocationUpdatesRule.pushLocationUpdate {
+                latitude = coordinates.first().latitude()
+                longitude = coordinates.first().longitude()
+            }
+            mockLocationReplayerRule.playRoute(route)
+
+            mapboxNavigation = MapboxNavigation(
+                NavigationOptions.Builder(activity)
+                    .accessToken(getMapboxAccessTokenFromResources(activity))
+                    .build()
+            )
+
+            mapboxNavigation.setRoutes(directionsResponse.routes())
+        }
+    }
+
+    @Test
+    fun selectAlternateRoute() {
+        addLocationPuck()
+        addRouteLine()
+        addNavigationCamera()
+
+        runOnMainSync {
+            assertEquals(3, routeLineApi.getRoutes().size)
+
+            val soonToBePrimaryRoute = routeLineApi.getRoutes()[1]
+            routeLineApi.updateToPrimaryRoute(soonToBePrimaryRoute)
+
+            assertEquals(soonToBePrimaryRoute, routeLineApi.getRoutes()[0])
+
+            mapboxNavigation.setRoutes(routeLineApi.getRoutes())
+            mapboxNavigation.registerRouteProgressObserver(object : RouteProgressObserver {
```

Comment:
`RouteProgressObserver` is currently not invoked, the test passes by finishing the test method but doesn't actually assert anything.

Reply:
If I set a break point at line: assertEquals(routeLineApi.getRoutes()[0], routeProgress.route) execution stops there.

Reply:
@cafesilencio that's right, but what you're seeing is a process crash, not an assertion failure. That's because the callback is delivered _after_ the test actually finishes.

Check out this simple diff. where effectively after the test finishes, I'm unregistering the listeners. It won't catch the issue anymore.
```diff
diff --git a/instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/ui/routeline/AlternativeRouteSelectionTest.kt b/instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/ui/routeline/AlternativeRouteSelectionTest.kt
index 5f6a0866c..2568ff94c 100644
--- a/instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/ui/routeline/AlternativeRouteSelectionTest.kt
+++ b/instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/ui/routeline/AlternativeRouteSelectionTest.kt
@@ -92,13 +92,15 @@ class AlternativeRouteSelectionTest : BaseTest<BasicNavigationViewActivity>(
                     // only need one route progress for this test
                     mapboxNavigation.unregisterRouteProgressObserver(this)
 
-                    assertEquals(routeLineApi.getRoutes()[0], routeProgress.route)
+                    assertEquals(routeLineApi.getRoutes()[1], routeProgress.route)
                 }
             })
             mapboxNavigation.startTripSession()
         }
 
         Espresso.onIdle()
+        // effectively end of test
+        mapboxNavigation.onDestroy() // unregisters listeners
     }
 
     fun addRouteLine() {

```

While in the correct setup, the ` Espresso.onIdle()` should hold us until all async operations finish. That's why we need an additional idling resource registered to actually hold the test execution until we make assertions in `RouteProgressObserver`. You can use the `RouteProgressStateIdlingResource` for that.

Reply:
Added the RouteProgressStateIdlingResource. I'm not sure it's in the right place but I tested by adding the changes you indicated above and I can see the test fail then pass when corrected.